lib/middleware/resource.js

```diff
@@ -1,47 +1,18 @@
 const debug = require('debug')('hydra-box:resource')
-const clownface = require('clownface')
-const ns = require('../namespaces')
-const { fromStream } = require('rdf-dataset-ext')
-const rdf = { ...require('@rdfjs/data-model'), ...require('@rdfjs/dataset') }
-const TermSet = require('@rdfjs/term-set')
 
-async function loadResource (store, term) {
-  const dataset = await fromStream(rdf.dataset(), store.match(null, null, null, term))
-
-  if (dataset.size === 0) {
-    return null
-  }
-
-  const types = new TermSet(clownface({ dataset, term }).out(ns.rdf.type).terms)
-
-  return {
-    term,
-    dataset,
-    types
-  }
-}
-
-async function findClassResource (store, term) {
-  return loadResource(store, term)
-}
-
-async function findPropertyResource (store, propertyTerm) {
-  const dataset = await fromStream(rdf.dataset(), store.match(null, null, propertyTerm, null))
-
-  if (dataset.size === 0) {
-    return null
-  }
+function factory ({ loader }) {
+  return async (req, res, next) => {
+    let resources = await loader.byClass(req.hydra.term)
 
-  const term = [...dataset][0].subject
+    if (resources.length === 0) {
+      resources = await loader.byProperty(req.hydra.term)
+    }
```

Comment:
My gut tells me that we should always have the option to check by property. Maybe does not make sense to interrogate the store right away.

Here's an algorithm I have in mind (using proposed names)

1. Find resources `bySubject`
2. Match supported operations
3. If a single operation is found, return
4. Find resources `byObject`
5. Match supported operations
6. If a single operation is found, return
7. If no operation was found, fail request

Reply:
The logic makes sense. Will be part of the operation finder PR.